src/pages/workspaces/hooks/useCloudEnvironmentPolling.ts

```diff
@@ -0,0 +1,98 @@
+  appDataDisks?: PersistentDisk[];
+}
+
+export const useCloudEnvironmentPolling = (
+  name: string,
+  namespace: string,
+  workspace?: Workspace
+): CloudEnvironmentDetails => {
+  const [controller, setController] = useState(new window.AbortController());
+  const abort = () => {
+    controller.abort();
+    setController(new window.AbortController());
+  };
+  const signal = controller.signal;
+
+  const timeout = useRef<NodeJS.Timeout>();
+  const [runtimes, setRuntimes] = useState<ListRuntimeItem[]>();
+  const [persistentDisks, setPersistentDisks] = useState<PersistentDisk[]>();
+  const [appDataDisks, setAppDataDisks] = useState<PersistentDisk[]>();
+
+  const saturnWorkspaceNamespace = workspace?.workspace?.namespace;
+  const saturnWorkspaceName = workspace?.workspace?.name;
+
+  const reschedule = (ms) => {
+    clearTimeout(timeout.current);
+    timeout.current = setTimeout(refreshRuntimesSilently, ms);
+  };
+  const load = async (maybeStale?: boolean): Promise<void> => {
+    try {
+      const cloudEnvFilters = _.pickBy((l) => !_.isUndefined(l), {
+        role: 'creator',
+        saturnWorkspaceName,
+        saturnWorkspaceNamespace,
+      }) as Record<string, string>; // we literally just filtered out the undefined values, but ts doesn't know this
+
+      // Disks.list API takes includeLabels to specify which labels to return in the response
+      // Runtimes.listV2 API always returns all labels for a runtime
+      const [newDisks, newRuntimes] = await Promise.all([
+        Ajax(signal)
+          .Disks.disksV1()
+          .list({
+            ...cloudEnvFilters,
+            includeLabels: 'saturnApplication,saturnWorkspaceName,saturnWorkspaceNamespace',
+          }),
+        Ajax(signal).Runtimes.listV2(cloudEnvFilters),
+      ]);
+
+      setRuntimes(newRuntimes);
+      setAppDataDisks(_.remove((disk) => _.isUndefined(getDiskAppType(disk)), newDisks));
+      setPersistentDisks(_.filter((disk) => _.isUndefined(getDiskAppType(disk)), newDisks));
+      const runtime = getCurrentRuntime(newRuntimes);
+      reschedule(
+        maybeStale ||
+          ['Creating', 'Starting', 'Stopping', 'Updating', 'LeoReconfiguring'].includes(
+            getConvertedRuntimeStatus(runtime) ?? ''
+          )
+          ? 10000
+          : 120000
+      );
+    } catch (error) {
+      reschedule(30000);
+      throw error;
+    }
+  };
+  const refreshRuntimes = withErrorReporting('Error loading cloud environments', load) as (
+    maybeStale?: boolean
+  ) => Promise<void>;
+  const refreshRuntimesSilently = withErrorIgnoring(load);
+  useEffect(() => {
+    if (
+      workspace?.workspaceInitialized &&
+      workspace.workspace.name === name &&
+      workspace.workspace.namespace === namespace
+    ) {
+      refreshRuntimes();
+    }
+    return () => {
+      clearTimeout(timeout.current);
+      abort();
+    };
+    //  eslint-disable-next-line react-hooks/exhaustive-deps
+  }, [name, namespace, workspace]);
```

Comment:
Since `name` and `namespace` are args to the hook, they can't change. `workspace` can change because it isn't initially initialized, and properties within it update.

Reply:
`name` and `namespace` _can_ change if the component is re-rendered with different props.

Reply:
But you pass them into `useCloudEnvironmentPolling` when you call the hook. Won't that always cause the `useEffect' to run because it's creating a new usage/instane of the hook? It's also quite possible that I still don't really understand the lifecycle of a hook function.

Reply:
@blakery's explanation in the other thread helped me understand this better. And it's true that one COULD call them with different props, we just don't because the name/namespace of a given workspace can't actually change. But having them listed as dependencies to `useEffect` makes sense.

Reply:
If Terra UI wasn't so dependent on using keys to work around `useOnMount`, navigating to the same page for a different workspace would possibly re-render the component with a difference namespace/name.